{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyMzBXzySIujwzKW84MzayBU",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/yongchanzzz/bioinformatics/blob/main/consensus_creator_colab.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# consensus_creator on Colab\n",
        "Consensus Creator from GitHub:\n",
        "https://github.com/TaizoAyase/consensus_creator"
      ],
      "metadata": {
        "id": "_P4dYoDm7eRb"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Import necessary libraries and upload files\n",
        "import re\n",
        "import csv\n",
        "from google.colab import files\n",
        "uploaded_files = files.upload()"
      ],
      "metadata": {
        "id": "WRTjpYgq7riW",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Choose parameters\n",
        "thresholds_input = \"0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1, 1.1, 1.2, 1.4, 1.6, 1.8, 2, 2.5, 3, 4\" #@param {type:\"string\"}\n",
        "\n",
        "def parse_thresholds(input_str):\n",
        "    # Split the input string by commas and convert each to a float\n",
        "    thresholds = [float(num.strip()) for num in input_str.split(',')]\n",
        "    return thresholds\n",
        "\n",
        "# Parse the input string to get a list of thresholds\n",
        "default_thresholds = parse_thresholds(thresholds_input)\n"
      ],
      "metadata": {
        "id": "sQkh9ZsU8GhZ",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "_uRAfDXf7D-g",
        "cellView": "form"
      },
      "outputs": [],
      "source": [
        "#@title Run code and downlaod\n",
        "\n",
        "#!/usr/bin/env python\n",
        "\"\"\"\n",
        "Consensus Creator ver 2.01 adapted to Google Colab\n",
        "\n",
        "This progam reads pre-aligned sequences as a single FASTA file, and creates\n",
        "consensus sequences. The program code was written by Yongchan Lee, 2017,\n",
        "and modified by Mizuki Takemoto. Adapted to Google Colab, 2024.\n",
        "Theoretical framework is based on the work by Steipe B et al., 1994.\n",
        "\n",
        "-Reference\n",
        "1) Steipe, B., Schiller, B., Pluckthun, A. & Steinbacher, S.\n",
        "Sequence statistics reliably predict stabilizing mutations in a protein\n",
        "domain. J. Mol. Biol. 240, 188-92 (1994).\n",
        "\n",
        "\n",
        "MIT License\n",
        "\n",
        "Copyright (c) 2020 Mizuki Takemoto\n",
        "\n",
        "Permission is hereby granted, free of charge, to any person obtaining a copy\n",
        "of this software and associated documentation files (the \"Software\"), to deal\n",
        "in the Software without restriction, including without limitation the rights\n",
        "to use, copy, modify, merge, publish, distribute, sublicense, and/or sell\n",
        "copies of the Software, and to permit persons to whom the Software is\n",
        "furnished to do so, subject to the following conditions:\n",
        "\n",
        "The above copyright notice and this permission notice shall be included in all\n",
        "copies or substantial portions of the Software.\n",
        "\n",
        "THE SOFTWARE IS PROVIDED \"AS IS\", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR\n",
        "IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,\n",
        "FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE\n",
        "AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER\n",
        "LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,\n",
        "OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE\n",
        "SOFTWARE.\n",
        "\"\"\"\n",
        "\n",
        "from __future__ import division, print_function\n",
        "from copy import deepcopy\n",
        "from datetime import datetime\n",
        "import numpy as np\n",
        "import sys\n",
        "\n",
        "class AminoAcidsCounter(object):\n",
        "    # TODO: key error if unusual letter is contained\n",
        "\n",
        "    all_aa_letters = set(\"ACDEFGHIKLMNPQRSTVWY-\")\n",
        "\n",
        "    def __init__(self, num):\n",
        "        self.id = num\n",
        "        self.dict = {}\n",
        "        for k in self.all_aa_letters:\n",
        "            self.dict[k] = 0\n",
        "\n",
        "    def calc_sum(self):\n",
        "        n_total = sum(self.dict.values())\n",
        "        self.n_gap = self.dict.pop(\"-\")\n",
        "        return n_total - self.n_gap\n",
        "\n",
        "    def get_max_aa(self):\n",
        "        return max([(v, k) for k, v in self.dict.items()])[1]\n",
        "\n",
        "\n",
        "class MultiFasta(object):\n",
        "    def __init__(self, filename):\n",
        "        self.filename = filename\n",
        "        self.fasta_ary = []\n",
        "        self.find_all_fasta(filename)\n",
        "\n",
        "    def __getitem__(self, i):\n",
        "        return self.fasta_ary[i]\n",
        "\n",
        "    def find_all_fasta(self, filename):\n",
        "        with open(filename, \"r\") as f:\n",
        "            line = f.readline().rstrip(\"\\r\\n\")\n",
        "\n",
        "            # the first line must be a header\n",
        "            if not re.match(\"^>\", line):\n",
        "                raise\n",
        "\n",
        "            header = re.sub(\"^>\", \"\", line)\n",
        "\n",
        "            line = f.readline().rstrip(\"\\r\\n\")\n",
        "            seq = \"\"\n",
        "\n",
        "            while line:\n",
        "                if re.match(\"^>\", line):\n",
        "                    self.fasta_ary.append(SingleFasta(header, seq))\n",
        "                    header = re.sub(\"^>\", \"\", line)\n",
        "                    seq = \"\"\n",
        "                else:\n",
        "                    seq += line\n",
        "                line = f.readline().rstrip(\"\\r\\n\")\n",
        "\n",
        "    def get_all(self, n):\n",
        "        return [fasta.seq[n] for fasta in self.fasta_ary]\n",
        "\n",
        "class SingleFasta(object):\n",
        "    def __init__(self, header, seq):\n",
        "        self.header = header\n",
        "        self.seq = self.clean_sequence(seq)\n",
        "\n",
        "    @staticmethod\n",
        "    def clean_sequence(seq):\n",
        "        valid_aa_letters = set(\"ACDEFGHIKLMNPQRSTVWY-\")\n",
        "        return ''.join(aa if aa in valid_aa_letters else '-' for aa in seq)\n",
        "\n",
        "    @property\n",
        "    def length(self):\n",
        "        return len(self.seq)\n",
        "\n",
        "d = datetime.now()\n",
        "timestamp_str = d.strftime('%Y%m%d_%H%M%S')\n",
        "output_file_path = f\"Consensus_{timestamp_str}.txt\"\n",
        "\n",
        "with open(output_file_path, 'w') as file:\n",
        "    sys.stdout = file\n",
        "    print(\"Imported file:\")\n",
        "    for filename in uploaded_files.keys():\n",
        "        print(filename)\n",
        "    print()\n",
        "    print(\"Threshold values:\")\n",
        "    print(\", \".join([str(e) for e in default_thresholds]))\n",
        "    print()\n",
        "sys.stdout = sys.__stdout__\n",
        "\n",
        "# Load Multi-Fasta file #\n",
        "for filename in uploaded_files.keys():\n",
        "    multifasta = MultiFasta(filename)\n",
        "    # proceed with your logic\n",
        "\n",
        "n_length = multifasta[0].length\n",
        "n_seq = len(multifasta.fasta_ary)\n",
        "\n",
        "# raise Error if all sequences do not have same length\n",
        "bool_ary = [fasta.length == n_length for fasta in multifasta.fasta_ary]\n",
        "if not all(bool_ary):\n",
        "    raise ValueError\n",
        "\n",
        "# make counts for each amino acids\n",
        "counts = {}\n",
        "for i in range(n_length):\n",
        "    counter = AminoAcidsCounter(i)\n",
        "    for aa in multifasta.get_all(i):\n",
        "        counter.dict[aa] += 1\n",
        "    counts[i] = counter\n",
        "\n",
        "# sum up over all sequences\n",
        "all_count = {}\n",
        "for i in range(n_length):\n",
        "    all_count[i] = counts[i].calc_sum()\n",
        "\n",
        "# header\n",
        "with open(output_file_path, 'a') as file:\n",
        "    sys.stdout = file\n",
        "    print(\"AA_Pos    WT  Freq    Con Freq    FCon/FWT\")\n",
        "\n",
        "seq_id = 1\n",
        "target_seq = []\n",
        "consensus_seq = []\n",
        "ratio_ary = []\n",
        "\n",
        "for i in range(n_length):\n",
        "    target_aa = multifasta[0].seq[i]\n",
        "\n",
        "    # skip gap\n",
        "    if target_aa == \"-\":\n",
        "        continue\n",
        "\n",
        "    target_freq = counts[i].dict[target_aa] / all_count[i] * 100\n",
        "    max_aa = counts[i].get_max_aa()\n",
        "    max_freq = counts[i].dict[max_aa] / all_count[i] * 100\n",
        "\n",
        "    ratio = max_freq / target_freq\n",
        "    ratio_ary.append(ratio)\n",
        "\n",
        "    # save amino acid sequences\n",
        "    target_seq.append(target_aa)\n",
        "    consensus_seq.append(max_aa)\n",
        "\n",
        "    output = (\n",
        "        \"{seq_id:<6d}{target_aa:>6s}  {target_freq:5.1f}    \"\n",
        "        \"{max_aa:<3s} {max_freq:5.1f}    {ratio:4.2f}\"\n",
        "    ).format(**locals())\n",
        "    # output = \"%s, %5.2f, %s, %5.2f\" % (target_aa, target_freq, max_aa, max_freq)\n",
        "    with open(output_file_path, 'a') as file:\n",
        "        sys.stdout = file\n",
        "        print(output)\n",
        "    sys.stdout = sys.__stdout__\n",
        "\n",
        "    seq_id += 1\n",
        "\n",
        "# output sequences\n",
        "title = multifasta[0].header\n",
        "with open(output_file_path, 'a') as file:\n",
        "    sys.stdout = file\n",
        "    print()\n",
        "    print(\">WT_\" + title)\n",
        "    print(\"\".join(target_seq))\n",
        "    print()\n",
        "    print(\">Consensus_\" + title)\n",
        "    print(\"\".join(consensus_seq))\n",
        "    print()\n",
        "sys.stdout = sys.__stdout__\n",
        "\n",
        "# print suggested mutation and seq\n",
        "target_seq = np.array(target_seq)\n",
        "consensus_seq = np.array(consensus_seq)\n",
        "ratio_ary = np.array(ratio_ary)\n",
        "\n",
        "for t in default_thresholds:\n",
        "    bool_ary = np.log(ratio_ary) > t\n",
        "\n",
        "    target = deepcopy(target_seq)\n",
        "\n",
        "    # mutation suggestion\n",
        "    base_seq = [\"-\" for i in range(seq_id - 1)]\n",
        "    base_seq = np.array(base_seq)\n",
        "    base_seq[bool_ary] = consensus_seq[bool_ary]\n",
        "\n",
        "    n_mutation = sum(bool_ary)\n",
        "    header = \">Threshold_value:{t} (suggested_mutations:{n_mutation})\".format(**locals())\n",
        "\n",
        "    with open(output_file_path, 'a') as file:\n",
        "        sys.stdout = file\n",
        "        print(header)\n",
        "        print(\"\".join(base_seq))\n",
        "        print()\n",
        "    sys.stdout = sys.__stdout__\n",
        "\n",
        "    # construct suggestion\n",
        "    target[bool_ary] = consensus_seq[bool_ary]\n",
        "    header = \">Construct_threshold_value:{}\".format(t)\n",
        "    with open(output_file_path, 'a') as file:\n",
        "        sys.stdout = file\n",
        "        print(header)\n",
        "        print(\"\".join(target))\n",
        "        print()\n",
        "    sys.stdout = sys.__stdout__\n",
        "\n",
        "print(f\"Data written to {output_file_path}\")\n",
        "files.download(output_file_path)"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Plot histogram\n",
        "# Plots for ln(Fcons/FWT)\n",
        "import numpy as np\n",
        "import matplotlib.pyplot as plt\n",
        "\n",
        "target_arr    = np.array(target_seq)\n",
        "consensus_arr = np.array(consensus_seq)\n",
        "ratio_arr     = np.array(ratio_ary, dtype=float)\n",
        "\n",
        "# Masks: finite and strictly >1\n",
        "finite_mask = np.isfinite(ratio_arr)\n",
        "valid_mask  = finite_mask & (ratio_arr > 1.0)\n",
        "\n",
        "ratios_gt1   = ratio_arr[valid_mask]\n",
        "ln_ratios_gt1 = np.log(ratios_gt1)\n",
        "\n",
        "if ratios_gt1.size == 0:\n",
        "    raise ValueError(\"No positions with Fcons/FWT > 1\")\n",
        "\n",
        "# Define bins\n",
        "nbins = 50 #@param {type:\"integer\"}\n",
        "edges_raw = np.linspace(ratios_gt1.min(), ratios_gt1.max(), nbins + 1)\n",
        "edges_ln  = np.linspace(ln_ratios_gt1.min(), ln_ratios_gt1.max(), nbins + 1)\n",
        "\n",
        "plt.figure()\n",
        "plt.hist(ln_ratios_gt1, bins=edges_ln)\n",
        "plt.xlabel(\"ln(Fcons/FWT)\")\n",
        "plt.ylabel(\"Number of residues\")\n",
        "plt.title(\"Histogram of ln(Fcons/FWT) (Fcons/FWT > 1)\")\n",
        "plt.tight_layout()\n",
        "plt.show()\n",
        "\n",
        "cons_diff = (consensus_arr != target_arr)\n",
        "thresholds_ln, counts_ln = [], []\n",
        "for t in edges_ln[:-1]:\n",
        "    mask = valid_mask & cons_diff & (np.log(ratio_arr) > t)\n",
        "    thresholds_ln.append(t)\n",
        "    counts_ln.append(int(mask.sum()))\n",
        "\n",
        "plt.figure()\n",
        "plt.plot(thresholds_ln, counts_ln, marker=\"o\")\n",
        "plt.xlabel(\"Threshold on ln(Fcons/FWT)\")\n",
        "plt.ylabel(\"Suggested mutations\")\n",
        "plt.title(\"Suggested mutations vs threshold\")\n",
        "plt.tight_layout()\n",
        "plt.show()"
      ],
      "metadata": {
        "cellView": "form",
        "id": "YFAQUXNRtX0Z"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}